bigint initial commit

td2r opened the pull request:
initial bigint pull request

Commit:
bigint initial commit

Commit:
fixed non-working operator < and >

kt-cpp commented on bigint/big_integer.cpp:
```diff
@@ -1,242 +1,436 @@
+
+    uint64_t const base = static_cast<uint64_t>(UINT32_MAX) + 1;
+
+    int compare(storage_t const &a, storage_t const &b) {
```
[note] этот `int` возвращает только  -1, 0 и 1. Тогда уж можно было бы сделать бул. Другой вариант - не делать лишние касты и, допустим, вместо `(>) - (<)` делать только `-`

td2r replied:
Пока писал ответ, осознал первый вариант...
Действительно, я использовал эту функцию там, где этот инт и не нужен. К тому же у меня уже есть похожая функция - smaller, на которую и переписал.

Второй вариант кажется нерабочим, если из 0 вычесть UINT32_MAX. Можно также убрать один каст:
вместо (a > b) - (a < b) написать 1 - ((a < b) << 1U), но мне кажется этот вариант уступает, поскольку использует два последовательных вычисления, а предыдущий - два параллельных и последовательное. К тому же касты от бул к инт не кажутся такими страшными, хотя точно в этом я не уверен.

kt-cpp commented on bigint/big_integer.cpp:
```diff
@@ -1,242 +1,436 @@
+    // return a % x
+    uint32_t divide(storage_t &a, uint32_t x) {
+        uint64_t carry = 0;
+        for (size_t i = a.size(); i != 0;) {
```
[note] выглядит ну такое себе, есть масса вариантов сделать информативнее
мне нравится следующий оч красивый трюк:
`for (size_t i = a.size() - 1; i-->0;)` 🍰

td2r replied:
Fixed. Взял на заметку.

kt-cpp commented on bigint/big_integer.cpp:
```diff
@@ -1,242 +1,436 @@
-{
-    mpz_init_set(mpz, other.mpz);
-}
+    uint32_t trial(__uint128_t r1, __uint128_t r2, __uint128_t r3, __uint128_t d2) {
```
[note] я лично люблю писать, что `__uint128_t` не определен в стандарте, а еще это gcc депендент. На досуге можешь подумать как не использовать тип ширины 128, но исправлять далеко не обязательно

td2r replied:
Да, мне тоже это не очень нравится, но Сорокин лично разрешил использовать 128. Возможно есть алгоритм без такой надобности, который я не знаю.
Есть вариант написать свое небольшое умножение и деление на двух 64 числах: либо как здесь
https://stackoverflow.com/questions/28868367/getting-the-high-part-of-64-bit-integer-multiplication
либо через asm вставки
и бинарить ответ для деления.
Производительность упадет, но код станет менее зависим от компилятора.
Было бы интересно услышать правильное решение.

kt-cpp replied:
asm-вставки сделают код зависимым не от компилятора, а от процессора :)

На самом деле для деления 128-битные типы (в том числе их имитации на двух 64-битных числах) вообще не обязательны — ты делишь 3 разряда на 2, а можно 2 на 1 — тогда достаточно 64-битных типов. В любом случае, такой вариант, хоть и зависим от компилятора, но сильно увеличивает производительность, поэтому исправлять не нужно.

kt-cpp commented on bigint/big_integer.h:
```diff
@@ -48,7 +50,15 @@ struct big_integer
+    void resize_(size_t n);
+
+private:
+    int sign_;
```
[minor] это лишние 3 байта на знак?) почему не `signed char`/`enum`?
С другой стороны можешь поаргументировать, я подумаю)

td2r replied:
enum точно не подойдет, поскольку я перемножаю/делю знаки. По памяти это скорее всего не даст никакого выигрыша, наплодятся static_cast, да и нет в нем особой надобности, интовое значение интуитивно понятно.
Скорее всего эти 3 байта останутся неиспользованными из-за struct padding. Мне нужен был какой-то аргумент со стороны стандарта, чтобы выбрать тип, ведь signed char создан для символов, а вкупе с предыдущим аргументом напрашивался самый простой вариант - int.
Но 3 байта это 3 байта, так что написал int8_t.

kt-cpp replied:
Здесь ты прав. Например, это можно проверить так: скомпилировать твой код, а потом [декомпилировать](https://cutter.re/) (или посмотреть в ассемблер, но там менее очевидно). Один из конструкторов bigint'а выглядит так:

![image](https://user-images.githubusercontent.com/67123994/86521383-8fb3f300-be69-11ea-9f41-f08f774ec6ea.png)

Здесь видно, что вектор появляется по адресу `arg1 + 8` — целых семь байт остаются лишними :) Можно заставить компилятор сэкономить эти байты, но не рекомендуется так делать.

kt-cpp reviewed (commented):
Очень хорошо, только notes & minor. Код -- *конфетка* 🍬 🥇 , еще и быстрый. Не жалко похвалить!

Commit:
bigint fix1
1. sign_ now int8_t instead of int
2. cycles kinda
   for (size_t i = n; i != 0;) {
       --i;
       ...
   }
   replaces with
   for (size_t i = n; i-->0;) {...}
3. changed sign formula from (a > b) - (a < b) to 1 - ((a < b) << 1U)

Commit:
bigint fix1
4. removed compare and replaced with less uses smaller
5. made reference symbol & related to type not to name

kt-cpp commented on bigint/big_integer.cpp:
```diff
@@ -1,242 +1,427 @@
-big_integer& big_integer::operator/=(big_integer const& rhs)
-{
-    mpz_tdiv_q(mpz, mpz, rhs.mpz);
+big_integer& big_integer::operator%=(const big_integer& rhs) {
```
[note] Кажется, в делении довольно дешево посчитать помимо частного ещё и остаток — было бы логично сделать функцию, которая возвращает пару, а потом в /= и %= отдавать нужные её элементы (например, так работает ассемблерная инструкция для деления `idiv`). Так ты избавляешься от тяжелого умножения, которое здесь происходит.
